src/dataflow/src/render/filter.rs

```diff
@@ -0,0 +1,273 @@
+    pub fn render_filter(
+        &mut self,
+        relation_expr: &MirRelationExpr,
+    ) -> (Collection<G, Row>, Collection<G, DataflowError>) {
+        if let MirRelationExpr::Filter { input, predicates } = relation_expr {
+            // Partition predicates into normal, and temporal lower/upper bounds.
+            let (normal, lower, upper) = extract_temporal(predicates.iter().cloned())
+                .unwrap_or_else(|err| {
+                    panic!("Temporal predicate error: {:?}", err);
+                });
+
+            let (ok_collection, err_collection) = self.collection(input).unwrap();
+
+            let (oks, errs) = ok_collection.inner.flat_map_fallible({
+                let mut datums = DatumVec::new();
+                move |(data, time, diff)| {
+                    let temp_storage = RowArena::new();
+                    let datums_local = datums.borrow_with(&data);
+                    // Ignore any records failing normal predicate evaluation.
+                    let mut passed = true;
+                    for pred in normal.iter() {
+                        passed = passed
+                            && match pred.eval(&datums_local, &temp_storage) {
+                                Err(e) => {
+                                    return Some(Err((DataflowError::from(e), time, diff)))
+                                        .into_iter()
+                                        .chain(None.into_iter());
+                                }
+                                Ok(Datum::True) => true,
+                                _ => false,
+                            }
+                    }
+                    if passed {
+                        // In order to work with times, it is easiest to convert it to an i128.
+                        // This is because our decimal type uses that representation, and going
+                        // from i128 to u64 is even more painful.
+                        let mut time_128 = i128::from(time);
+                        // Track whether we have seen a null in either bound, as this should
+                        // prevent the record from being produced at any time.
+                        let mut null_eval = false;
+
+                        // If there is a lower bound, advance `time` to that lower bound.
+                        // This decimal stuff is brittle; let's hope the scale never changes.
+                        for l in lower.iter() {
+                            match l.eval(&datums_local, &temp_storage) {
+                                Err(e) => {
+                                    return Some(Err((DataflowError::from(e), time, diff)))
+                                        .into_iter()
+                                        .chain(None.into_iter());
+                                }
+                                Ok(Datum::Decimal(s)) => {
+                                    if time_128 < s.as_i128() {
```

Comment:
it seems like we are specifically advancing to the maximum lower bound right? could maybe call that out in the comments a bit more I had to reread a few times before I could convince myself this was predicate order independent